### PR TITLE
Add Developer Applications Demo using Transformers Library

### PR DESCRIPTION
**Pull request by hupreti**
Developer Applications on Cloud AI 100 using Transformers Library

app_ui

### CODE REVIEW

File: app/cert.pem [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+WHSTYX1YTHmJ9TtvjZES5qCDrpXbZpcOnW7qbxehRSrzaaIaKWEevFcQXVUhA5wI
+JbfZs/kXu5eGxzqg95YyjHlLk2lbGQYz7hlkjgM5D4ekALCbqLWFuNP6QqsSC/bi
+tS2Sim1NVRwFuWgTa5V0t53RvNZkN75m1bPlQkOW4ROlifP9zsRHQA09UaHpRqUC
+VueqorrOUOkpWcZ3e0YuPWAk/4xljZI4iw75XaSwiZ/BrUXDCNnHr+KLaUuI+AOI
+9io2njAp7aDx8zLIU9m7fdTBnN6qvfCsL5zoI2hfzOrz+Ir8ValruwYmWH/hV/0P
+xaT+zufZTqFEQsPCgXp4z0r+fVeUYwcAUnjD65/IPXbYvs/42LEkVZZjamPdohZk
+nmiOObbX+a8637ANq8FRdn6VlpXNdkq1cWGn65xXSNbfFG6xcIV7w+oH03jEyWx1
+wtT9Rr+xVlf9vTgO2Ao9UoKpImVU9wiXmRQLZeGrez8S8YVJhLN6HaKIV3S9uS7U
+ug0DpsSLpd1xDhGwzLEpDRUCAwEAAaNTMFEwHQYDVR0OBBYEFJoVccLyUC8VHude
+akBh7kNF9qX/MB8GA1UdIwQYMBaAFJoVccLyUC8VHudeakBh7kNF9qX/MA8GA1Ud
```
Comment (vbaddi) [inline app/cert.pem:22]: is it okay to put this `cert.pem` here,? we can avoid this.

Comment (hupreti) [reply app/cert.pem:22]: Sure, will remove this and add instructions in readme to generate them

File: app/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+gradio==4.29
```
Comment (vbaddi) [inline app/requirements.txt:1]: we can add this in the main setup.py? or we want to enable to route of application development in different requirements.txt?

Comment (hupreti) [reply app/requirements.txt:1]: depends on the type of approval we get, since we are not keeping the developed app inside the pip package, it doesn't make sense to add this in setup.py

File: QEfficient/generation/LLMGenerator.py [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+
+class LLMGenerator:
+    def __init__(
+        self,
+        qpc_path,
+        model_name,
+        device_id: Optional[List[int]] = [0],
+        prompt_len: Optional[int] = 32,
+        ctx_len: Optional[int] = 128,
+        streamer: Optional["BaseStreamer"] = None,
```
Comment (ochougul) [inline QEfficient/generation/LLMGenerator.py:107]: Why is this type hinted like this?

File: QEfficient/generation/LLMGenerator.py [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+        logits_processor: Optional = None,
```
Comment (ochougul) [inline QEfficient/generation/LLMGenerator.py:108]: Wrong type hint
```suggestion
        logits_processor: Optional[class] = None,
```

**Review by ochougul — CHANGES_REQUESTED** [review]

File: app/utils.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# -----------------------------------------------------------------------------
+#
+# Copyright (c)  2023-2024 Qualcomm Innovation Center, Inc. All rights reserved.
```
Comment (quic-jhugo) [inline app/utils.py:3]: Please double check the year range here

File: app/Readme.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+#
+# ----------------------------------------------------------------------------- -->
+# Developer Applications on Cloud AI 100 using Transformers Library 
+
+
+### Instructions to launch the app 
+1. System Dependencies 
+    - `sudo apt-get install ffmpeg openssl`
+    - same as the `efficient-transformers`
+2. Clone the repo `git clone https://github.com/hupreti/efficient-transformers.git`
```
[the remark this answers is not in the document]
Comment (hupreti) [reply app/Readme.md:14]: thanks will update

File: app/Readme.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+3. Change directory `cd app`
+  - create `app_config.json` inside directory
+  - update the information in app_config.json like given below section
+4. Create a python3.8 environment `python3.8 -m venv env` and activate the environment using `source env/bin/activate`
```
[the remark this answers is not in the document]
Comment (hupreti) [reply app/Readme.md:18]: we are not distributing any qeff env, are we? we are only distributing qeff python package

File: QEfficient/generation/LLMGenerator.py [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+from threading import Thread
+from typing import *
+import torch
+import numpy as np
+import torch.nn as nn
+
+import transformers
+
+
+from QEfficient.generation.aic_infer import QAICInferenceSession
```
[the remark this answers is not in the document]
Comment (hupreti) [reply QEfficient/generation/LLMGenerator.py:25]: thanks, will update

File: app/utils.py (hunk 5 above)
Comment (hupreti) [reply app/utils.py:3]: year ranges same are kept same

File: QEfficient/generation/LLMGenerator.py (hunk 3 above)
Comment (hupreti) [reply QEfficient/generation/LLMGenerator.py:107]: fixed

File: QEfficient/generation/LLMGenerator.py [hunk 9]
```diff
@@ -0,0 +1,478 @@
...
+
+def print_to_string(*args, **kwargs):
+    output = io.StringIO()
+    print(*args, file=output, **kwargs)
+    contents = output.getvalue()
+    output.close()
+    return contents
+
+
+def write_io_files(
```
[the remark this answers is not in the document]
Comment (hupreti) [reply QEfficient/generation/LLMGenerator.py:61]: removed

File: QEfficient/generation/LLMGenerator.py [hunk 10]
```diff
@@ -0,0 +1,478 @@
...
+        else:
+            inputs["position_ids"] += 1
+
+        batch_size, prompt_len = inputs["input_ids"].shape
+        return inputs, prompt_len
+
+    def generate(self, prompt: str, sample: bool = False, max_new_tokens: int = None):
+        session = self.session
+        # if self.write_io_dir:
+        #     write_io_files(inputs, outputs, write_io_dir, "prefill")
```
[the remark this answers is not in the document]
Comment (hupreti) [reply QEfficient/generation/LLMGenerator.py:308]: removed

File: QEfficient/generation/LLMGenerator.py [hunk 11]
```diff
@@ -0,0 +1,478 @@
...
+    LogitsProcessorList,
+    MinLengthLogitsProcessor,
+    TopKLogitsWarper,
+    TemperatureLogitsWarper,
+    StoppingCriteriaList,
+    MaxLengthCriteria,
+)
+
+
+def print_to_string(*args, **kwargs):
```
[the remark this answers is not in the document]
Comment (hupreti) [reply QEfficient/generation/LLMGenerator.py:53]: removed

File: app/utils.py (hunk 5 above)
Comment (quic-jhugo) [reply app/utils.py:3]: What does "year ranges are kept same" mean?
Lets ask a specific question - when exactly did you create this file?

Comment (hupreti) [reply app/utils.py]: fixed in latest commit

File: QEfficient/generation/llm_generator.py [hunk 12]
```diff
@@ -0,0 +1,298 @@
+# -----------------------------------------------------------------------------
+#
+# Copyright (c)  2024 Qualcomm Innovation Center, Inc. All rights reserved.
+# SPDX-License-Identifier: BSD-3-Clause
+#
+# -----------------------------------------------------------------------------
+
+import json
```
Comment (ochougul) [inline QEfficient/generation/llm_generator.py:8]: remove unused

File: QEfficient/generation/llm_generator.py [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+import os
+import torch
+import numpy as np
+import torch.nn as nn
+
+from typing import Dict, List, Optional, Union
```
Comment (ochougul) [inline QEfficient/generation/llm_generator.py:14]: Dict unused

File: QEfficient/generation/llm_generator.py [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+from threading import Thread
+
+from QEfficient.generation.cloud_infer import QAICInferenceSession
```
Comment (ochougul) [inline QEfficient/generation/llm_generator.py:17]: this import should be below transformers import.
read more here: https://peps.python.org/pep-0008/#imports

File: QEfficient/generation/llm_generator.py [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+        prompt_len: Optional[int] = 32,
+        ctx_len: Optional[int] = 128,
+        streamer: Optional[Union[TextStreamer, TextIteratorStreamer]] = None,
+        logits_processor: Optional = None,
+        logits_warper: Optional = None,
+    ):
+        self.session = None
+        self.tokenizer = None
+        self.is_first_prompt = False
+        self.model_name = ""
```
Comment (ochougul) [inline QEfficient/generation/llm_generator.py:49]: Can be replaced with line 66, is there a purpose of assigning empty string here?

File: QEfficient/generation/llm_generator.py [hunk 16]
```diff
@@ -0,0 +1,298 @@
...
+        self.qpc_path = ""
```
Comment (ochougul) [inline QEfficient/generation/llm_generator.py:50]: replace with 61-63?

File: QEfficient/generation/llm_generator.py [hunk 17]
```diff
@@ -0,0 +1,298 @@
...
+        self.device_id = [0]
```
Comment (ochougul) [inline QEfficient/generation/llm_generator.py:51]: replace with line 64

File: QEfficient/generation/llm_generator.py [hunk 18]
```diff
@@ -0,0 +1,298 @@
...
+class LLMGenerator:
+    def __init__(
+        self,
+        qpc_path,
+        model_name,
+        device_id: Optional[List[int]] = [0],
+        prompt_len: Optional[int] = 32,
+        ctx_len: Optional[int] = 128,
+        streamer: Optional[Union[TextStreamer, TextIteratorStreamer]] = None,
+        logits_processor: Optional = None,
```
Comment (ochougul) [inline QEfficient/generation/llm_generator.py:43]: both `logits_processor` nd `logits_warper` are unused.

File: QEfficient/generation/llm_generator.py [hunk 19]
```diff
@@ -0,0 +1,298 @@
...
+    ):
+        self.session = None
+        self.tokenizer = None
+        self.is_first_prompt = False
+        self.model_name = ""
+        self.qpc_path = ""
+        self.device_id = [0]
+        self.curr_cache_index = 0
+        self.ctx_len = ctx_len
+        self.retained_state = True
```
Comment (ochougul) [inline QEfficient/generation/llm_generator.py:54]: Looks like this code currently supports only if `self.retained_state=True`, if not please make this variable a function parameter, and assign based on that.
Also, please test when `self.retained_state=False`, I suppose this is the case when we run the model `BertStyle`.

File: QEfficient/generation/llm_generator.py [hunk 20]
```diff
@@ -0,0 +1,298 @@
...
+        self.write_io_dir = False
```
Comment (ochougul) [inline QEfficient/generation/llm_generator.py:55]: Unused?

File: app/app.py [hunk 21]
```diff
@@ -0,0 +1,309 @@
...
+    global last_prompt, previous_current_ctx_len, last_state_generation_ids
+
+    qeff_generator_model.curr_cache_index = 0
+    qeff_generator_model.generated_ids = []
+
+    if qeff_generator_model.curr_cache_index >= qeff_generator_model.ctx_len - 1:
+        qeff_generator_model.curr_cache_index = 0
+
+    qeff_generator_model.curr_cache_index = 0
+    qeff_generator_model.stop_indicator = True
```
Comment (ochougul) [inline app/app.py:151]: remove, default value is True.

File: app/app.py [hunk 22]
```diff
@@ -0,0 +1,309 @@
...
+
+
+def summary():
+    return summary_text
+
+
+def infer_prompt(msg, chat_history, task, model):
+    global last_prompt, previous_current_ctx_len, last_state_generation_ids
+
+    qeff_generator_model.curr_cache_index = 0
```
Comment (ochougul) [inline app/app.py:144]: is this required? default is 0 anyway

File: app/app.py [hunk 23]
```diff
@@ -0,0 +1,309 @@
...
+    qeff_generator_model.generated_ids = []
+
+    if qeff_generator_model.curr_cache_index >= qeff_generator_model.ctx_len - 1:
+        qeff_generator_model.curr_cache_index = 0
```
Comment (ochougul) [inline app/app.py:148]: Why is this required?

File: app/app.py [hunk 24]
```diff
@@ -0,0 +1,309 @@
...
+
+    qeff_generator_model.curr_cache_index = 0
```
Comment (ochougul) [inline app/app.py:150]: ?

File: QEfficient/generation/llm_generator.py [hunk 25]
```diff
@@ -0,0 +1,298 @@
...
+            tokenizer = AutoTokenizer.from_pretrained(
+                model_name, padding_side="left", hf_token=hf_token
+            )
+            if tokenizer.pad_token_id is None:
+                tokenizer.pad_token_id = tokenizer.eos_token_id
+            self.tokenizer = tokenizer
+        except Exception as err:
+            raise RuntimeError(f"Unable to load tokenizer, {err}")
+
+        if streamer:
```
Comment (ochougul) [inline QEfficient/generation/llm_generator.py:100]: handle else case, raise error

File: QEfficient/generation/llm_generator.py [hunk 26]
```diff
@@ -0,0 +1,298 @@
...
+                )
+
+        except Exception as err:
+            raise RuntimeError(f"Unable to load qpc on device , {err}")
+
+        try:
+            hf_token = None
+            if os.getenv("HF_TOKEN") is not None:
+                hf_token = os.getenv('HF_TOKEN')
+            tokenizer = AutoTokenizer.from_pretrained(
```
Comment (ochougul) [inline QEfficient/generation/llm_generator.py:91]: Have you tested this on models present in tests? I have seen it failing with config file not found on huggingface, and the work-around is to use `hf_download`, check code in `QEfficient/cloud/infer.py::80-87` you can allow only tokenizer file pattern, as you don't need other model files for tokenizer

**Review by ochougul — COMMENTED** [review]
The review is not yet complete. Will review the rest later.

File: QEfficient/cloud/infer.py [hunk 27]
```diff
@@ -60,6 +60,7 @@ def main(
     device_group: List[int] = [
         0,
     ],
+    execute : bool = True
```
Comment (ochougul) [inline QEfficient/cloud/infer.py:63]: this funtionality exists in `QEfficinet/cloud/execute.py::main`, not required to add here.

File: app/app.py [hunk 28]
```diff
@@ -0,0 +1,300 @@
...
+            time.sleep(0.0005)
+        summary_text += f"$ Downloaded {model_name} from cache directory\n"
+        # TODO : call QEfficient compile api
+        progress(0, desc="Optimizing and Compiling...")
+        time.sleep(0.5)
+        for i in progress.tqdm(range(100), desc="Optimizing and Compiling..."):
+            time.sleep(0.07)
+
+        # calling infer api directly to get qpc_path
+        app_config[task][model_name]['qpc_path'] = infer_api(
```
[the remark this answers is not in the document]
Comment (ochougul) [reply app/app.py:121]: If you want you can create a utils function called `compile_and_execute` and put it inside, `QEffcieint/cloud/utils.py`
and then change infer API as well to use same.

It's your choice.

File: QEfficient/cloud/infer.py [hunk 29]
```diff
@@ -60,8 +60,10 @@ def main(
     device_group: List[int] = [
         0,
     ],
+    skip_stats : bool = False,
```
Comment (ochougul) [inline QEfficient/cloud/infer.py:63]: @vbaddi @hupreti 
Can we please make a new function called `export_and_compile` inside infer API that returns `qpc_path` and has the logic of not regenerating the `qpc/ONNX` if they already exist, 

This way:
 - `infer.py::main` can use this function
 - as well as app,
 - as well as tests that we are planning
 
We can meet if you guys want me to explain this.  Please let me know if you guys have a better solution.

Comment (hupreti) [reply QEfficient/cloud/infer.py:63]: restored cli interface, we need to have a single interface which will perform all things for user so it is not just compile and execute

File: QEfficient/cloud/infer.py [hunk 30]
```diff
@@ -76,21 +77,24 @@ def main(
     onnx_dir_path = os.path.join(model_card_dir, "onnx")
     onnx_model_path = os.path.join(onnx_dir_path, model_name.replace("/", "_") + "_kv_clipped_fp16.onnx")
 
+    # skip model download if qpc exits and we do not need stats
+    if not qpc_exists(qpc_dir_path) or not skip_stats:
```
Comment (ochougul) [inline QEfficient/cloud/infer.py:81]: As we remove `skip_stats`, this code can be moved to line 123.

File: QEfficient/cloud/infer.py (hunk 29 above)
Comment (ochougul) [reply QEfficient/cloud/infer.py:63]: Can we move the `latency_kv_stats` call in the main function and remove it from `infer_api`
get `qpc path`, `model_hf_path` from `infer_api` prepare tokenizer using `model_hf_path` and pass `tokenizer` and `qpc_path`  to `latency_kv_stats` inside main function.

I agree that we need single interface to perform all, that will be main function in the file.

And that way, I think you don't have to use any extra parameter. You can also rename `infer_api` to `export_and_compile` then.

File: QEfficient/cloud/infer.py [hunk 31]
```diff
@@ -76,21 +77,24 @@ def main(
...
     # Get tokenizer
-    if hf_token is not None:
-        login(hf_token)
-    model_hf_path = hf_download(
-        repo_id=model_name,
-        cache_dir=cache_dir,
-        ignore_patterns=["*.txt", "*.onnx", "*.ot", "*.md", "*.tflite", "*.pdf"],
-    )
-    tokenizer = AutoTokenizer.from_pretrained(model_hf_path, use_cache=True, padding_side="left")
+        if hf_token is not None:
```
Comment (ochougul) [inline QEfficient/cloud/infer.py:83]: This code (line 83-90) can also be moved under main function.
and pass `model_hf_path` as input to `infer_api`, and return `qpc_path` as output.

Comment (hupreti) [reply QEfficient/cloud/infer.py:83]: why should user downloads hf_model files from model_card and provide you model_hf_path ? please keep in mind that this is high_level api for user

File: QEfficient/generation/llm_generator.py (hunk 26 above)
Comment (hupreti) [reply QEfficient/generation/llm_generator.py:83]: yes, hf_download are used un-necessarily, even at place where we just need tokenizer still we are downloading all files @ochougul please raise an issue

File: app/app.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (hupreti) [reply app/app.py:117]: @anujgupt-github @vbaddi I believe we need discuss on requirements on high-level apis and low level apis, their use-cases and design principles. @ochougul please take https://github.com/daquexian/onnx-simplifier as example to understand top-level cli api and top-level programmable api

File: QEfficient/generation/llm_generator.py (hunk 26 above)
Comment (ochougul) [reply QEfficient/generation/llm_generator.py:83]: Here, we download all the files while only needing tokenizer, assuming that user must already have all the files in `cache_dir`, in which case, the files won't be downloaded again, as this was how this API is supposed to be used.

If you want to change it to only download tokenizer files when only tokenizer is needed, you are welcome to update it. This is definitely a better way to do this.

You can raise issue internally yourself, no need to ask anybody. Thanks.

Comment (hupreti) [reply QEfficient/generation/llm_generator.py:83]: we cannot assume user will have these files downloaded, IMO just bad design, so didn't move out line 80-87 to main, it will be required in top level api. Consider scaling this solution as aws service and user just wants use qeff apis to run inference on some edge device or device with low space, how will your solution pan out? Happy to discuss more.

**Comment by anujgupt-github** [comment]
Not planned to be revived at the moment, will revisit if really needed. 
Qualcomm has a much more elaborate developer applications platform which uses efficient-transformers, and can be leveraged
